### ISSUE TITLE
Implement `filter` query param on `listImages`

### ISSUE THREAD

**Issue by johnflavin** [spotify/docker-client#348]
The `listImages` method is missing a feature in the Docker Remote API: filtering by image name, a.k.a. repo tag.

From the [Remote API docs](https://docs.docker.com/reference/api/docker_remote_api_v1.22/#list-images) (latest API version, though this quoted section has not changed in any recent API version):

> * filters – a JSON encoded value of the filters (a map[string][]string) to process on the images list. Available filters:
 `dangling=true`
 `label=key` or `label="key=value"` of an image label

> * filter - only return images with the specified name

Currently, `listImages` only supports the former. All `ListImagesParam`s passed into the method get appended to the request url as values for the `filters` query param. There is no way to give any value for the `filter` query param, thus no way to filter the full list of images by image name.

(See #180 for a question that appears to have this issue at its root. It looks like @bwahn is trying to query for images by name, but the `listImages` method does not allow it.) 

I think that supporting the `filter` query param and allowing `listImages` to be filtered by image name would be a useful feature. But is there a reason that it should not be implemented? On its face this feature seems to duplicate functionality: one can already use `inspectImage(name)` to query about a particular tagged image, so why support `listImages` with `filter=name`? The reason is that those two methods return different information. The former returns an `ImageInfo` and the latter a `List` of `Image`s. The information contained in either an `Image` or `ImageInfo` object doesn't completely overlap the other. (Specifically, `ImageInfo` almost contains a superset of the fields of `Image`, except that `ImageInfo` does not contain the image's `Labels`. For some reason Docker doesn't report those on `GET /images/(name)/json`, which is the request made by `inspectImage`.)

**Comment by mattnworb** [comment]
> I think that supporting the filter query param and allowing listImages to be filtered by image name would be a useful feature. But is there a reason that it should not be implemented? 

I agree it would be useful. I think we'd be happy to merge a PR implementing this functionality.

**Comment by bwahn** [comment]
@johnflavin Thanks for detailed that issue. Using latest version of Docker, docker-client of spotify, I'm going to test it.

**Comment by johnflavin** [comment]
I did some digging, and I was totally wrong. This is already supported.

Here's why...
`listImages` takes an array of `ListImagesParam`s. Those objects are pretty much just a key and a value. Some of those params may be instances of the subclass `ListImagesFilterParam`. Only those instances of the subclass will be sent through to the request as `filters=["key1":"value1","key2":"value2",...]`. Everything else will be sent through as `key=value`.

So, in short, to filter by image name you just need to call
`listImages(new ListImagesParam("filter", imageName))`.

(On a side note, in figuring out that the feature I want already exists, I discovered that some of the image label code I wrote for PR #341 does not play nicely with other filter params. But I'll open a new issue for that.)

**Comment by davidxia** [comment]
Yea, our docs are lacking a bit on describing all the new features that
have been added recently...

On Tue, Feb 16, 2016 at 4:36 PM, Flavin <notifications@github.com> wrote:

> I did some digging, and I was totally wrong. This is already supported.
>
> Here's why...
> listImages takes an array of ListImagesParams. Those objects are pretty
> much just a key and a value. Some of those params may be instances of the
> subclass ListImagesFilterParam. Only those instances of the subclass will
> be sent through to the request as
> filters=["key1":"value1","key2":"value2",...]. Everything else will be
> sent through as key=value.
>
> So, in short, to filter by image name you just need to call
> listImages(new ListImagesParam("filter", imageName)).
>
> (On a side note, in figuring out that the feature I want already exists, I
> discovered that some of the image label code I wrote for PR #341
> <https://github.com/spotify/docker-client/pull/341> does not play nicely
> with other filter params. But I'll open a new issue for that.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/issues/348#issuecomment-184883858>
> .
>



-- 
David Xia
Software Engineer
dxia@spotify.com
